clmock/clmock.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2023 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package clmock
+
+import (
+	"context"
+	"time"
+	"github.com/ethereum/go-ethereum/beacon/engine"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type CLMock struct {
+	ctx context.Context
+	cancel context.CancelFunc
+}
+
+// Start invokes the clmock life-cycle function in a goroutine
+func (c *CLMock) Start() {
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+	go c.clmockLoop()
+}
+
+// Stop halts the clmock service
+func (c * CLMock) Stop() {
+	c.cancel()
+}
+
+// TODO: use ctx with timeout when calling rpc methods? is there a way they could hang indefinitely (even though we are calling on same machine/process)?
+
+// clmockLoop manages the lifecycle of clmock.
+// it drives block production, taking the role of a CL client and interacting with Geth via the engine API
+func (c *CLMock) clmockLoop() {
+	ticker := time.NewTicker(time.Millisecond * 500)
+	blockPeriod := time.Second * 2 // hard-coded fast block period for testing purposes
+	lastBlockTime := time.Now()
+
+	var curForkchoiceState *engine.ForkchoiceStateV1
+	var prevRandaoVal common.Hash
+	var suggestedFeeRecipient common.Address
+
+	engine_api := engineAPI{}
+	if err := engine_api.Connect(c.ctx, "http://127.0.0.1:8545"); err != nil {
+		log.Error("failed to connect to engine api: %v", err)
+	}
+
+	header, err := engine_api.GetHeaderByNumber(c.ctx, 0)
+	if err != nil {
+		log.Error("failed to get genesis block header", err)
+	}
+
+	curForkchoiceState = &engine.ForkchoiceStateV1{
+		HeadBlockHash: header.Hash(),
+		SafeBlockHash: header.Hash(),
+		FinalizedBlockHash: header.Hash(),
+	}
+
+	_, err = engine_api.ForkchoiceUpdatedV1(c.ctx, curForkchoiceState, nil)
+
+	if err != nil {
+		log.Error("failed to initiate PoS transition for genesis via Forkchoiceupdated", err)
+	}
+
+	for {
+		select {
+		case _ = <-c.ctx.Done():
+			break
+		case curTime := <-ticker.C:
+			if curTime.After(lastBlockTime.Add(blockPeriod)) {
+				safeHead, err := engine_api.GetHeaderByTag(c.ctx, "safe")
```

Comment:
I feel like these "safe" and "finalized" fetches are not needed. There's no other proposer so it should be able to keep track of these blocks (they are already tracked in `curForkchoiceState`)?